### PR TITLE
Wrap package in a project directory

### PR DESCRIPTION
**Pull request by rgreinho**
This patch wraps the package generated by the `build_yaml` or
`build_json` functions in a project directory instead of writing the
files directly in the destination directory.

The project directory takes the same name as the package.

Drive-by:
* Updates `.gitignore`.

### CODE REVIEW

File: open_alchemy/build/__init__.py
Comment (jdkandersson) [inline open_alchemy/build/__init__.py:247]: Do you think it would be better to move the decision about which folder to store the package in to the `dump` function? It is designed to do everything it needs to for output, already has the `name` parameter and has logic to deal with files (such as looking for folders and creating them if necessary).

**Review by jdkandersson — CHANGES_REQUESTED** [review]
Thanks for the pull request!

- [ ] There are some type errors I believe tied to how the dump function is called (it would now receive a path object instead of a string). This would be solved by the comment I put on that file.

There are also 2 tests that are failing which are integration tests for the build function:

- [ ] `tests/open_alchemy/test_build.py::test_execute`
- [ ] `tests/open_alchemy/integration/test_init.py::test_build_json`

You can see more details by looking at the pipeline output. I think they should be straight forward to fix, they are just looking for the old folder structure.

- [ ] I would also recommend adding some tests on the `dump` function where a directory for the output does and does not exist

File: open_alchemy/build/__init__.py
Comment (rgreinho) [reply open_alchemy/build/__init__.py]: Oh you're right. That makes complete sense to update the path closer to the logic. I will change that.

**Comment by rgreinho** [comment]
I moved this PR to draft because I am experimenting and I am editing some test settings that will not be edited in the final version.

`python-bandit` seems to be causing some trouble.

**Comment by rgreinho** [comment]
I got it to work, but I had to use some trickery:

* if I leave `python-bandit`, `pytest` fails with the following error:
```bash
INTERNALERROR> Failed: Direct construction of BanditItem has been deprecated, please use BanditItem.from_parent.
INTERNALERROR> See https://docs.pytest.org/en/stable/deprecations.html#node-construction-changed-to-node-from-parent for more details.
```
* I could not get the coverage to 100% without using a trick in the `dump()` function and replacing:
```python
if not pathlib_package_path.is_dir():
        pathlib_package_path.mkdir()
```
with
```python
pathlib_package_path.mkdir(exist_ok=True)
```
It does the same thing, but it still bugs me that I could not get the other way to get covered. The test `test_dump_path_not_exists()` is supposed to cover this case right?

**Comment by rgreinho** [comment]
I opened an issue for the `pytest-bandit` plugin: https://github.com/Wanderu/pytest-bandit/issues/18, to ask whether the plugin is supposed to work with pytest 6. 

It may be worth investigating using `bandit` directly.

**Review by jdkandersson — COMMENTED** [review]
> I got it to work, but I had to use some trickery:
> 
> * if I leave `python-bandit`, `pytest` fails with the following error:
> 
> ```shell
> INTERNALERROR> Failed: Direct construction of BanditItem has been deprecated, please use BanditItem.from_parent.
> INTERNALERROR> See https://docs.pytest.org/en/stable/deprecations.html#node-construction-changed-to-node-from-parent for more details.
> ```
> 
> * I could not get the coverage to 100% without using a trick in the `dump()` function and replacing:
> 
> ```python
> if not pathlib_package_path.is_dir():
>         pathlib_package_path.mkdir()
> ```
> 
> with
> 
> ```python
> pathlib_package_path.mkdir(exist_ok=True)
> ```
> 
> It does the same thing, but it still bugs me that I could not get the other way to get covered. The test `test_dump_path_not_exists()` is supposed to cover this case right?

I don't mind that the logic is changed, in fact, I think this is nicer and easier to read in the new way. I think `test_dump_path_name_exists` and `test_dump_path_name_is_file` already test for whether `dist/<name>` exist or are files, since a new `<name>` layer is being added we would likely need `dist/<name>/<name>` folder existing and another test where `dist/<name>/<name>` is a file. It would be cool if these could be parametrised with pytest? That would be a stretch goal :wink:

**Comment by jdkandersson** [comment]
> I opened an issue for the `pytest-bandit` plugin: [Wanderu/pytest-bandit#18](https://github.com/Wanderu/pytest-bandit/issues/18), to ask whether the plugin is supposed to work with pytest 6.
> 
> It may be worth investigating using `bandit` directly.

I think removing `pytest-bandit` from `pytest` is ok, maybe we can add it into the Azure pipelines instead? It should be straight forward, take a look at the `azure-pipelines.yml` file for an example. You would likely have to add `bandit` as a dev dependency and add something like the `black` job.

File: open_alchemy/build/__init__.py
Comment (jdkandersson) [inline open_alchemy/build/__init__.py:209]: Thats excellent!

**Comment by jdkandersson** [comment]
That error `bandit` is throwing, I understand where it is coming from. I don't mind if you leave it in, it isn't part of the intent for this pull request and was there the whole time. Once this is merged that error can be fixed as another issue.

**Comment by rgreinho** [comment]
I ended up skipping all the bandit warnings: `B101`, `B303`, `B310` and opened an other issue to be sure we fix them at some point (https://github.com/jdkandersson/OpenAlchemy/issues/191).

**Review by jdkandersson — APPROVED** [review]

**Comment by jdkandersson** [comment]
I'm wondering, if I mark this as approved, can you hit the merge button?

**Comment by rgreinho** [comment]
Nope, it is only available to contributors with write access. 

However if you want to automate this process, I would recommend using Kodiak (https://kodiakhq.com/). It is really good! And helped me streamlining the review/rebase/merge process a lot!

**Pull request merged**
